### DISCUSSION TITLE
How to properly type include statement for nested relations

### DISCUSSION THREAD

**Discussion by abmenzel** [prisma/prisma#21957] [Q&A] [answered]
### Question

I am trying to create an include statement as such

```typescript
	const map = await prisma.map.findUnique({
		where: {
			id: Number(req.query.mapId),
		},
		include: {
			...detailedMapFields,
		},
	})
```

Where`detailedMapFields` refer to following

```typescript
export const detailedMapFields: Prisma.MapInclude = {
	owner: true,
	participants: {
		include: {
			user: true,
			territories: true,
		},
	},
	invites: {
		include: {
			to: true,
		},
		where: {
			status: 'PENDING',
		},
	},
	events: {
		include: {
			author: true,
			involved: true,
			map: true,
		},
		orderBy: {
			createdAt: 'desc',
		},
	},
}
```

I am doing this to ensure that i'm always including the same fields in the different queries I make.
My issue here is that typescript doesn't infer the that territories is included on participants

```typescript
	map.participants.forEach((p) => {
		console.log(p.territories) // Typescript doesn't know that territories is included
	})
```

I suppose I have to extend the `Prisma.MapInclude` type, but I can't find any documentation on how to do it. What is the proper way of typing this?

**Comment by kenzeasy** [comment, marked as answer]
Hey @abmenzel I got a similar usecase where I want to include a specific set after my queries.

From what I could test [prisma validator](https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety/prisma-validator) seems to do the trick.

I have tested with something similar to your example but not as deeply nested as you and it seems to be ok:
```ts
import { Prisma } from "@prisma/client"
export const detailedMapFields = Prisma.validator<Prisma.MapInclude>()({
    owner:true,
    //...
    //all the things you want to include ...
    //...
})
```
Then you can use your object as you did or without the spread operator (both seems to work on my side)
```ts
const map = await prisma.map.findUnique({
    where: {
        id: Number(req.query.mapId),
    },
    include: {
        ...detailedMapFields,
    },
    //or
    //include:detailedMapFields
})
```
From the little I did to reproduce your usecase typescript seems aware of everything I have put in the `detailedMapFields`

Let me know if this worked for you

**Reply by abmenzel** [reply]
@kenzeasy yes! That did the trick, exactly what I was looking for :)
I wasn't aware of prisma-validator, thanks for pointing that out.
